Add LiteX SPI driver

mateusz-holenko opened the pull request:
This adds LiteX SPI driver and enables it in litex_vexriscv board.

lpawelcz committed:
dts: litex: add LiteX SoC builder vendor

Add LiteX SoC builder vendor with prefix "litex".

Signed-off-by: Pawel Czarnecki <pczarnecki@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

craviee committed:
boards: litex_vexriscv: Enable LiteX SPI driver

This commit enables LiteX SPI driver for litex_vexriscv board.

Signed-off-by: Daniel Craviee <dcraviee@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

craviee committed:
drivers: spi: Add LiteX SPI driver

This commit adds LiteX SPI drivers and its bindings.

Signed-off-by: Daniel Craviee <dcraviee@internships.antmicro.com>
Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko commented:
Rebased to resolve conflicts after merging #18649.

mateusz-holenko commented:
@tbursztyka Could you take a look at this PR?

tbursztyka commented on drivers/spi/spi_litespi.h:
```diff
@@ -0,0 +1,71 @@
+#define SPI_MOSI_DATA_ADDR (SPI_BASE_ADDR + 0x0c)
+#define SPI_MISO_DATA_ADDR (SPI_BASE_ADDR + 0x10)
+#define SPI_CS_ADDR (SPI_BASE_ADDR + 0x14)
+#define SPI_LOOPBACK_ADDR (SPI_BASE_ADDR + 0x18)
```
so these above are registers? Maybe better mentioning it in the name (with _REG added?)

mateusz-holenko replied:
Replaced `_ADDR` suffix with `_REG` suffix, e.g. `SPI_BASE_REG` for clarity.

tbursztyka commented on drivers/spi/spi_litespi.h:
```diff
@@ -0,0 +1,71 @@
+{
+	csr_writel(value >> 8, addr);
+	csr_writel(value, addr + 0x4);
+}
```
looks like riscv arch is missing sys_io.h implementation. Could you do that? it will certainly be useful for other part of the system on that architecture. And you would not have to re-implement it here for instance.

mateusz-holenko replied:
We could definiately use `sys_read8`/`sys_write8` from `common/sys_io.h` and drop `csr_writel`/`csr_readl`/`read8`/`write8`.

As of `write16`, the situation is a bit more complicated. The way we handle 16-bit registers here is not RISC-V specific, but rather LiteX specific. LiteX is an SoC builder that can be configured in many ways (which influences the way `write16` should behave - it depends on `data_width` and `bus_width` parameters) and that can target different CPUs (so this driver is not limited to RISC-V cores). As a result I'm not sure if `sys_io.h` is the place to put this code.

I do agree, though, that it would be good to have a `litex_common.h` header that provides implementation of `read16`/`write16`/`read32`/`write32` functions for all LiteX drivers. Do you have a suggestion where should such header be put? Is simply `include/litex_common.h` (or `include/litex.h`) a proper place?

mateusz-holenko replied:
We could definiately use `sys_read8`/`sys_write8` from `common/sys_io.h` and drop `csr_writel`/`csr_readl`/`read8`/`write8`.

As of `write16`, the situation is a bit more complicated. The way we handle 16-bit registers here is not RISC-V specific, but rather LiteX specific. LiteX is an SoC builder that can be configured in many ways (which influences the way `write16` should behave - it depends on `data_width` and `bus_width` parameters) and that can target different CPUs (so this driver is not limited to RISC-V cores). As a result I'm not sure if `sys_io.h` is the place to put this code.

I do agree, though, that it would be good to have a `litex_common.h` header that provides implementation of `read16`/`write16`/`read32`/`write32` functions for all LiteX drivers. Do you have a suggestion where should such header be put? Is simply `include/litex_common.h` (or just `include/litex.h`) a proper place?

tbursztyka replied:
include/ is for generic headers. Instead, I would embed the specific functions into soc/riscv/litex-vexriscv/soc.h

mateusz-holenko replied:
Sure. For now it's good enough to put it into `soc/riscv/litex-vexriscv`. In the future, if more LiteX-based platforms are added this should be moved to a more generic location.

tbursztyka commented on drivers/spi/spi_litespi.c:
```diff
@@ -0,0 +1,173 @@
+static struct spi_driver_api spi_litespi_api = {
+	.transceive = spi_litespi_transceive,
+	.release = spi_litespi_release,
+};
```
you are missing the async function if CONFIG_SPI_ASYNC if set: just add them and make them return -ENOTSUP.

mateusz-holenko replied:
Added an empty implementation of `spi_litespi_transceive_async`.

tbursztyka commented on drivers/spi/spi_litespi.c:
```diff
@@ -0,0 +1,173 @@
+		return -EBUSY;
+	}
+	return 0;
+}
```
so your driver does not handle the lock_on attribute of spi_config?

mateusz-holenko replied:
No it doesn't support this, in `spi_config` we verify that `SPI_LOCK_ON` is *not* set.

tbursztyka reviewed (changes requested):
rework the style a bit

mateusz-holenko committed:
soc: riscv: litex: Add helpers for accessing CSRs

Depending on LiteX configuration, CSRs
(control&status registers) might be split
into several consecutive registers.

This introduces common helper functions
for all LiteX drivers providing access
to CSRs for a default LiteX configuration
(data_width = 8bit, bus_width = 32bit).

Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko commented:
@tbursztyka Thanks for the review! I addressed your comments. Does it look good now?

MaureenHelm commented on dts/bindings/spi/litex,spi0.yaml:
```diff
@@ -0,0 +1,18 @@
+description: >
+    This binding gives a base representation of the LiteX SPI
+
+compatible: "litex,spi0"
```
Why is this `spi0` instead of `spi`? This looks like an instance number, which shouldn't be in the compatible

mateusz-holenko replied:
I changed compatible to "litex,spi" and renamed the whole file to "litex,spi.yaml"

MaureenHelm reviewed (changes requested):
Please add `spi` to the supported list in `boards/riscv/litex_vexriscv/litex_vexriscv.yaml` and make sure `tests/drivers/spi/spi_loopback` passes

mateusz-holenko committed:
samples: spi_fujitsu_fram: fix dependencies

This sample will not compile for boards
not supporting GPIO.

Signed-off-by: Mateusz Holenko <mholenko@antmicro.com>

mateusz-holenko commented:
@MaureenHelm Adding `spi` to supported list caused `spi_fujitsu_fram` sample to fail - I had to add a new commit fixing dependencies list for this sample. After the fix all tests pass.

I confirm that `tests/drivers/spi/spi_loopback` passes on HW.
